perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java

```diff
@@ -3923,42 +3999,106 @@ private void updateExistingGroupsWhileSynchronization(PerunSession sess, Group b
 	 * @param groupToUpdate for which will be parent specified
 	 * @param baseGroup under which may be new parent
 	 * @param candidateGroup with potential new parent group
+	 * @param loginAttributeName attribute name for login of group
 	 * @return updated parent group (if null, parent group hasn't changed)
 	 * @throws InternalErrorException
 	 */
-	private Group specifyParentForUpdatedGroup(PerunSession sess, Group groupToUpdate, Group baseGroup, CandidateGroup candidateGroup) throws InternalErrorException {
+	private Group specifyParentForUpdatedGroup(PerunSession sess, Group groupToUpdate, Group baseGroup, CandidateGroup candidateGroup, String loginAttributeName) throws InternalErrorException {
+		//get current parent of updated group, if there is no group (null), it means the group is under the base group
+		String oldParentGroupLogin = getOldParentGroupWithLogin(sess, groupToUpdate, baseGroup, loginAttributeName);
 
-		String actualParentName;
-		try {
-			actualParentName = getGroupById(sess, groupToUpdate.getParentGroupId()).getShortName();
-		} catch (GroupNotExistsException e) {
-			throw new InternalErrorException(e);
+		//check if there is difference between parents of current and new group, if not, return null
+		if(Objects.equals(oldParentGroupLogin, candidateGroup.getParentGroupLogin())) return null;
+
+		//if there is difference, we need to find a new parent group
+		Group newParentGroup = getNewParentGroupWithLogin(sess, candidateGroup, baseGroup, loginAttributeName);
+
+		//null in parent group means base group, return it
+		if(newParentGroup == null) {
```

Comment:
I think this can happen only if the baseGroup is null. Therefore, I think it is not necessary.

Reply:
Also right, removed.